zerver/tests/test_message_delete.py

```diff
@@ -321,3 +322,27 @@ def test_delete_message_in_unsubscribed_private_stream(self) -> None:
         result = self.client_delete(f"/json/messages/{msg_id}")
         self.assert_json_success(result)
         self.assertFalse(Message.objects.filter(id=msg_id).exists())
+
+    def test_update_first_message_id_on_stream_message_deletion(self) -> None:
+        realm = get_realm("zulip")
+        stream_name = "test"
+        cordelia = self.example_user("cordelia")
+        self.make_stream(stream_name)
+        self.subscribe(cordelia, stream_name)
+        message_ids = [self.send_stream_message(cordelia, stream_name) for _ in range(5)]
+        first_message_id = message_ids[0]
+
+        message = Message.objects.get(id=message_ids[3])
+        do_delete_messages(realm, [message])
+        stream = get_stream(stream_name, realm)
+        self.assertEqual(stream.first_message_id, first_message_id)
+
+        first_message = Message.objects.get(id=first_message_id)
+        do_delete_messages(realm, [first_message])
+        stream = get_stream(stream_name, realm)
+        self.assertEqual(stream.first_message_id, message_ids[1])
+
+        all_messages = Message.objects.filter(id__in=message_ids)
+        do_delete_messages(realm, all_messages)
+        stream = get_stream(stream_name, realm)
+        self.assertEqual(stream.first_message_id, None)
```

Comment:
Doing at least one `self.assert_database_query_count` check here would be helpful.